libyul/optimiser/Semantics.cpp

```diff
@@ -125,21 +125,24 @@ map<YulString, SideEffects> SideEffectsPropagator::sideEffects(
 	{
 		YulString funName = call.first;
 		SideEffects sideEffects;
-		util::BreadthFirstSearch<YulString>{call.second, {funName}}.run(
-			[&](YulString _function, auto&& _addChild) {
-				if (sideEffects == SideEffects::worst())
-					return;
-				if (BuiltinFunction const* f = _dialect.builtin(_function))
-					sideEffects += f->sideEffects;
-				else
-				{
-					if (ret.count(_function))
-						sideEffects += ret[_function];
-					for (YulString callee: _directCallGraph.functionCalls.at(_function))
-						_addChild(callee);
-				}
+		auto _visit = [&, visited = std::set<YulString>{}](YulString _function, auto&& _recurse) mutable {
+			if (visited.count(_function))
+				return;
+			visited.insert(_function);
```

Comment:
```suggestion
			if (!visited.insert(_function).second))
				return;
```
would actually be the same and would be faster - but it's a readability/speed tradeoff.

Reply:
Changed